packages/datagrid/src/selectionmodel.ts

```diff
@@ -0,0 +1,337 @@
+
+  /**
+   * A signal emitted when the selection model has changed.
+   */
+  get changed(): ISignal<this, void> {
+    return this._changed;
+  }
+
+  /**
+   * The data model associated with the selection model.
+   */
+  readonly model: DataModel;
+
+  /**
+   * Get the selection model for the model.
+   */
+  get selectionMode(): SelectionModel.SelectionMode {
+    return this._selectionMode;
+  }
+
+  /**
+   * Set the selection mode for the model.
+   *
+   * #### Notes
+   * This will clear the selection model.
+   */
+  set selectionMode(value: SelectionModel.SelectionMode) {
+    // Bail early if the mode does not change.
+    if (this._selectionMode === value) {
+      return;
+    }
+
+    // Update the internal mode.
+    this._selectionMode = value;
+
+    // Clear the current selections.
+    this.clear();
+  }
+
+  /**
+   * Test whether any selection intersects a row.
+   *
+   * @param index - The row index of interest.
+   *
+   * @returns Whether any selection intersects the row.
+   *
+   * #### Notes
+   * This method may be reimplemented in a subclass.
+   */
+  isRowSelected(index: number): boolean {
+    return some(this.selections(), s => Private.containsRow(s, index));
+  }
+
+  /**
+   * Test whether any selection intersects a column.
+   *
+   * @param index - The column index of interest.
+   *
+   * @returns Whether any selection intersects the column.
+   *
+   * #### Notes
+   * This method may be reimplemented in a subclass.
+   */
+  isColumnSelected(index: number): boolean {
+    return some(this.selections(), s => Private.containsColumn(s, index));
+  }
+
+  /**
+   * Test whether any selection intersects a cell.
+   *
+   * @param row - The row index of interest.
+   *
+   * @param column - The column index of interest.
+   *
+   * @returns Whether any selection intersects the cell.
+   *
+   * #### Notes
+   * This method may be reimplemented in a subclass.
+   */
+  isCellSelected(row: number, column: number): boolean {
+    return some(this.selections(), s => Private.containsCell(s, row, column));
+  }
+
+  /**
+   * Emit the `changed` signal for the selection model.
+   *
+   * #### Notes
+   * Subclasses should call this method whenever the selection model
+   * has changed so that attached data grids can update themselves.
+   */
+  protected emitChanged(): void {
+    this._changed.emit(undefined);
+  }
+
+  private _changed = new Signal<this, void>(this);
+  private _selectionMode: SelectionModel.SelectionMode = 'cell';
+}
+
+
+/**
+ * The namespace for the `SelectionModel` class statics.
+ */
+export
+namespace SelectionModel {
+  /**
+   * A type alias for the selection mode.
+   */
+  export
+  type SelectionMode = 'row' | 'column' | 'cell';
+
+  /**
+   * A type alias for the clear mode.
+   */
+  export
+  type ClearMode = 'all' | 'current' | 'none';
+
+  /**
+   * A type alias for the select args.
+   */
+  export
+  type SelectArgs = {
```

Comment:
Is this explicitly *not* `type SelectArgs = Selection & { ... }` for clarity's sake?

Reply:
More or less. There is no intrinsic link between the args and a selection (the model is free to modify the selection as it sees fit), so I didn't want to imply one in the typing.